### PR TITLE
Instant Search: Update prototype with downstream changes

### PR DESCRIPTION
**Pull request by jsnmoon**
## Changes proposed in this Pull Request:
Please note that all changes in this PR relate to Jetpack Instant Search. Specifically, this PR:
* Adds result filtering and sorting
* Adds pagination with support for infinite scrolling
* Enables searching from any site page with a sidebar widget area
* Polishes search result styling
* Fixes theme incompatibilities
* Adds caching to API requests
* Adds TrainTracks analytics

<details> 
<summary>This PR is the culmination of many downstream PRs. Click here to reveal downstream PR ID's.</summary>
#13371 #13393, #13430, #13505, #13377, #13545, #13605, #13614, #13602, #13684, #13714, #13721, #13750, #13748, #13713, #13742, and #13730
</details>

### Is this a new feature or does it add/remove features to an existing part of Jetpack?
Yes, this is part of the Jetpack Instant Search project: p3QzjZ-V3-p2. After landing this in master, we plan on presenting this work to our partners.

#### Testing instructions:
* Add `define( "JETPACK_SEARCH_PROTOTYPE", true );` to your wp-config.php. 
If using Jetpack's Docker development environment, you can create a file at `/docker/mu-plugins/instant-search.php` and add the define there.

* Ensure that your site has the Jetpack Pro plan and Jetpack Search enabled. 
You can enable Jetpack Search in the Performance tab within the Jetpack menu (`/wp-admin/admin.php?page=jetpack#/performance`).

* Select a theme of your choosing and add a Jetpack Search widget to the site via the customizer, preferably with some filters enabled. If you're using a theme with a sidebar widget area, please add the Jetpack Search widget there.
> Screen Shot 2019-10-15 at 4 14 41 PM

* If using a theme with a sidebar widget, you can navigate to `/` to start the search experience. Otherwise, navigate to your search page (e.g. `/?s=hello`).

* Try interacting with the search interface via the input field or the filter checkboxes. 

> Screen Shot 2019-10-15 at 4 15 41 PM

* Ensure that the interface works as expected. If you have any questions, please let us know!

#### Proposed changelog entry for your changes:
Launches a prototype Instant Search and Filtering intended to eventually replace the current Jetpack Search.

### CODE REVIEW

**Comment by gibrown** [comment]
The demo site is also running the `instant-search-master` branch if anyone just wants to play with the interaction and try out some different themes: http://gibrown.wpsandbox.me/demo/

File: modules/search/class.jetpack-search.php
Comment (jeherve) [inline modules/search/class.jetpack-search.php]: I am curious about this. Does this mean we'll now track and log searches from any visitor, as soon as the Jetpack master owner has connected the site to WordPress.com and thus implicitly accepted our ToS? Is that alright? What happens if that master owner has opted out using the Privacy settings at `wp-admin/admin.php?page=jetpack#/privacy`?

In relation with the GDPR, what happens if a specific user requests that all their search logs be deleted? Is that something that the site owner would have to comply to, and how would they go about this?

File: modules/search/instant-search/components/search-results.jsx
Comment (jeherve) [inline modules/search/instant-search/components/search-results.jsx]: Do we need to add the textdomain when using `_n` here and below?
https://github.com/WordPress/gutenberg/tree/master/packages/i18n#usage

File: modules/search/instant-search/components/search-app.jsx
Comment (jeherve) [inline modules/search/instant-search/components/search-app.jsx]: This doesn't quite work for me:
https://jeremy.hu/?s=marvel

```
Uncaught TypeError: Cannot read property 'lastChild' of null
    at removeChildren (dom.js:2)
    at search-app.jsx:59
```

I do have a widget in the footer though.

**Review by jeherve — COMMENTED** [review]
I wanted to test this on my own personal site, but I'm not having any luck. I'm leaving `JETPACK_SEARCH_PROTOTYPE` on there so you can play a bit with it. The search widget is in the footer.
https://jeremy.hu/?s=marvel&orderby=relevance&order=DESC&post_type=post%2Cpage%2Ctraktivity_event

*********

As another note, I am a bit worried that the first thing folks will want to do is customize the markup, and there is currently no easy to do this.

*********
I should also note that the bundle is currently too big to be shipped to production with the current limits. The production build currently errors out:

```
Message:
    asset size limit: The following asset(s) exceed the recommended size limit (34.2 KiB).
This can impact web performance.
Assets:
  jp-search.bundle.js (104 KiB)
```

**Comment by jsnmoon** [comment]
> I wanted to test this on my own personal site, but I'm not having any luck. I'm leaving `JETPACK_SEARCH_PROTOTYPE` on there so you can play a bit with it. The search widget is in the footer.
https://jeremy.hu/?s=marvel&orderby=relevance&order=DESC&post_type=post%2Cpage%2Ctraktivity_event @jeherve 

I pushed out a fix for this bug. Feel free to give it another shot!

> I should also note that the bundle is currently too big to be shipped to production with the current limits. The production build currently errors out

I'll look into this.

**Comment by jeherve** [comment]
Thanks! I am playing with it some more, and purposefully playing with the filters to see if they behave as expected. When searching for a term and applying one of the post type filters, I expected all posts that do not belong to that post type to be removed, but I still see `post` post types among the results:
https://jeremy.hu/?s=marvel&orderby=relevance&order=DESC&post_type=traktivity_event

In some cases, the whole searchform seems to disappear, alongside my options to clear filters / launch another search:
https://jeremy.hu/?s=marvel&orderby=relevance&order=DESC&post_type=post%2Cpage%2Ctraktivity_event&trakt_show=humans

**Comment by jsnmoon** [comment]
@jeherve: I noticed Instant Search doesn't inject properly onto your site; it's currently failing to replace the original filter checkboxes with the Instant Search checkboxes.

I think this is happening due to a mismatch between the widget ID provided by the server (`jetpack-search-filters-2` via `window.JetpackInstantSearchOptions`) and the one rendered in the DOM ('jetpack-search-filters-2-wrapper`). Can you help me better understand what's going on here?

File: modules/search/class.jetpack-search.php
Comment (jsnmoon) [reply modules/search/class.jetpack-search.php]: > What happens if that master owner has opted out using the Privacy settings at `wp-admin/admin.php?page=jetpack#/privacy`?

For those reading along, the Privacy Setting has a toggle control with the following text:

> Share information with our analytics tool about your use of services while logged in to your WordPress.com account.

In all honesty, I feel uneasy with our current implementation. Yes, the toggle there exists to satisfy the letter of the Cookie Law. Since Instant Search (ideally) won't handle user data via cookies, we technically won't need to respect the setting there.

However, it's important to note that we will begin tracking *all* visitors to any Jetpack site with Instant Search enabled with zero recourse for opting out.

I think we really ought to respect this site option for user analytics in Instant Search.

@bperson, your thoughts on this would be appreciated.

File: modules/search/instant-search/components/search-result-minimal.jsx
Comment (jeherve) [inline modules/search/instant-search/components/search-result-minimal.jsx]: Instead of relying on only the locale to display the date, do you think we could pass the date format from `get_option( 'date_format' )` as well, to get exactly the format used on the site, that may not be the default one for that locale.

As an example, my site's locale is `en_US`, but the date format I specified under Settings > General is `F j, Y`, not the short `m/d/y` version currently picked up from the locale and `dateStyle` right now.

**Review by jeherve — COMMENTED** [review]
> This should be fixed! This was happening due to a mismatch between the widget ID provided by the server and the one rendered in the DOM.

It does seem to work better now indeed! 
image

It also surfaces some interesting issues, where the numbers next to those filters don't all seem to be correct. I do not have 766 posts tagged with `duplicate content`, I have only one. :)

It is also interesting that the filters now appear by default on my home page, even before I launch a search. That wasn't the case before.

Since the filters are there, I can use them without entering any search term, and that returns results that do not really make sense to me, I don't really understand how those posts are ranked here:
https://jeremy.hu/?post_types=traktivity_event

************

It would be nice if the infinite loader used a similar style to the one used for Infinite Scroll module, to keep things consistent.

*******

Some of the Gridicons appear to be missing right now:

image

*******

I seem to be missing some of the filters I set up in the widget. Only the 2 first filters are listed right now:

image

*******

If I am looking at a post, and then use the widget to launch a search, the search term is appended to the post URL, like so:
https://jeremy.hu/music-marvel-cinematic-universe/?s=gyver

That's problematic since loading that URL directly won't work. I think the search term should be appended to the home URL, not the URL of the page you're on.

*********

When I start searching when on a post page, I don't get many tips that the main content area is now displaying my search results:

https://cloudup.com/cv67TEtx00k

I realize this may not be an issue with all themes, but I think it could be a problem with themes with footer widgets for example, especially on smaller screens.

*******

If I hit my browser's return key after launching a search, the search query remains, but with one less character in the term I typed. I think the whole search query should disappear, don't you think?

**Comment by jsnmoon** [comment]
> Some of the Gridicons appear to be missing right now:

This should be fixed! I was being overzealous when pruning the Gridicons earlier this week.

> If I am looking at a post, and then use the widget to launch a search, the search term is appended to the post URL, like so:
jeremy.hu/music-marvel-cinematic-universe/?s=gyver
>
> That's problematic since loading that URL directly won't work. I think the search term should be appended to the home URL, not the URL of the page you're on.

This should also be fixed, good catch!

> When I start searching when on a post page, I don't get many tips that the main content area is now displaying my search results:
> I realize this may not be an issue with all themes, but I think it could be a problem with themes with footer widgets for example, especially on smaller screens.

This is a definite weakness of the current design, which primarily serves themes with sidebar widget areas. I'll go ahead and note this in #13027.

File: modules/widgets/search.php
Comment (jsnmoon) [reply modules/widgets/search.php]: Good suggestion! I went ahead and replaced `defined` calls with `Constants::is_true( 'JETPACK_SEARCH_PROTOTYPE' )`.

**Comment by jsnmoon** [comment]
> It also surfaces some interesting issues, where the numbers next to those filters don't all seem to be correct. I do not have 766 posts tagged with duplicate content, I have only one. :)

cc @gibrown 

> It is also interesting that the filters now appear by default on my home page, even before I launch a search. That wasn't the case before.
> 
> Since the filters are there, I can use them without entering any search term, and that returns results that do not really make sense to me, I don't really understand how those posts are ranked here:
jeremy.hu/?post_types=traktivity_event

Populating filters before searching is a new behavior we're introducing with Instant Search. I think @gibrown can expand upon how the posts are initially ranked.

> I seem to be missing some of the filters I set up in the widget. Only the 2 first filters are listed right now:

It looks like we're only returning search results for `Tags` and `Types` at the moment. The other three taxonomy filters are returning zero results -- cc @gibrown.

> It would be nice if the infinite loader used a similar style to the one used for Infinite Scroll module, to keep things consistent.

I'll note this in the master issue thread.

File: modules/search/instant-search/components/search-result-minimal.jsx
Comment (jsnmoon) [reply modules/search/instant-search/components/search-result-minimal.jsx]: I think it'll be tricky given that we don't load any date formatting libraries for the sake of minimizing our bundle size. I'll look into this and report back.

Comment (bluefuton) [reply modules/search/instant-search/components/search-result-minimal.jsx]: Can we do the date formatting on the API side instead?

**Comment by gibrown** [comment]
> It also surfaces some interesting issues, where the numbers next to those filters don't all seem to be correct. I do not have 766 posts tagged with duplicate content, I have only one. :)

I just reindexed all content and I still see 766 posts. I suspect that you need to run a full bulk sync to fix it. Decided to leave that to you for now so you can check it out and verify that makes sense.

> It is also interesting that the filters now appear by default on my home page, even before I launch a search. That wasn't the case before.

Yep, new feature that was requested by LiquidWeb. Makes us more competitive with pure filtering based plugins also. It should also make customization easier since the end user will see the filters they are configuring even if they are not doing a search in the customizer.

> I don't really understand how those posts are ranked here: jeremy.hu/?post_types=traktivity_event

Hmmm, ya I guess it is using the default sorting when first selected. @jsnmoon can you force the default to be reverse chronological if there is no search query? I think that would be a more intuitive experience for now. I think ideally we should have popularity based sorting, but the current sorting doesn't do as good of a job of that yet.

> I seem to be missing some of the filters I set up in the widget. Only the 2 first filters are listed right now:

I suspect this is related to syncing problems. The index doesn't have the show taxonomy in it for instance, but I think that it should be getting indexed.

> It would be nice if the infinite loader used a similar style to the one used for Infinite Scroll module, to keep things consistent.

Punting on this probably makes the most sense for the prototype, but does this try to match the existing style of the theme?

**Comment by gibrown** [comment]
> I realize this may not be an issue with all themes, but I think it could be a problem with themes with footer widgets for example, especially on smaller screens.

@jeherve ya the "right" solution to this is that for mobile devices or themes that do not have a sidebar, we should create a search box that gets inserted at the top of the results area and scroll the page to that location. Similarly you have a problem with this theme that when you use the search box in the header then for some reason scrolling the page is disabled.

We weren't able to work on mobile for this release, and I think we should explicitly acknowledge that as a known issue in the release. Also maybe acknowledge that many of the same issues exist for themes that don't have a sidebar. This reminds me that I should start writing up the propotype launch announcement...

**Comment by gibrown** [comment]
@jsnmoon the change to have the query string not get appended to the end of post urls breaks sites that have a more complicated url structure. For instance http://gibrown.wpsandbox.me/demo/

We may want to separate some of these issues out and then fix them in a separate PR just to make sure we get a bit more testing on them. I think the filter checkboxes are also broken a bit. I click them and then the filter gets applied, but the checkbox is no longer checked.

**Comment by jsnmoon** [comment]
> the change to have the query string not get appended to the end of post urls breaks sites that have a more complicated url structure. For instance http://gibrown.wpsandbox.me/demo

This should now be fixed.

**Comment by jsnmoon** [comment]
> I think the filter checkboxes are also broken a bit. I click them and then the filter gets applied, but the checkbox is no longer checked.

This should also be fixed!

**Comment by jsnmoon** [comment]
> @jsnmoon can you force the default to be reverse chronological if there is no search query? I think that would be a more intuitive experience for now. 

This should be fixed!

File: modules/search/instant-search/components/search-result-minimal.jsx
Comment (jeherve) [reply modules/search/instant-search/components/search-result-minimal.jsx]: If we can't do much right now, could we at least move away from this terrible `m/d/y` format :P Anything would be better. :)

**Review by jeherve — COMMENTED** [review]
> I suspect that you need to run a full bulk sync to fix it. Decided to leave that to you for now so you can check it out and verify that makes sense.

Unfortunately that didn't seem to help :/

*********

Something seems off with the way tags are displayed now:
https://jeremy.hu/?post_tag=genericons

image

**********

Can you try scrolling down the page and searching for something on my site?
https://jeremy.hu/

I get a weird behaviour right now that didn't happen before:

http://cl.ly/7cc67b80f28f

***********

The bundle size still seems to be an issue:

```
AssetsOverSizeLimitWarning in plugin "webpack"
Message:
    asset size limit: The following asset(s) exceed the recommended size limit (60 KiB).
This can impact web performance.
Assets:
  jp-search.bundle.js (60.7 KiB)
Details:
    details: undefined
    missing: undefined
    origin: undefined
    dependencies: undefined
    module: undefined
    assets: [object Object]
```

**********

All filters sometimes disappear when you pick a different filter:
http://cl.ly/fdb60ce2b72a

It stops you from being able to filter things down even more.

***********

> Would it help if we separate some of these issues out and then fix them in a separate PR just to make sure we get a bit more testing on them? @jeherve what is your sense of where we are at with this merge? It feels like we're moving along well and it may make sense to treat some things as bugs rather than blockers?

Yes, definitely agree. Let's try to merge this and address what comes from now on in new PRs imo.

**Comment by jsnmoon** [comment]
> Something seems off with the way tags are displayed now:
jeremy.hu/?post_tag=genericons

This should be fixed.

> Can you try scrolling down the page and searching for something on my site?
jeremy.hu

I'm not able to reproduce this on Chrome 77, unfortunately. 

> The bundle size still seems to be an issue

:/ I'll go ahead and bump it to 70KiB.

> All filters sometimes disappear when you pick a different filter:
http://cl.ly/fdb60ce2b72a
>
> It stops you from being able to filter things down even more.

This usually means we've run out of things to filter on; perhaps @gibrown can pitch in on this one.

**Comment by gibrown** [comment]
> All filters sometimes disappear when you pick a different filter ... It stops you from being able to filter things down even more.
> This usually means we've run out of things to filter on; perhaps @gibrown can pitch in on this one.

Ya, so our filtering is done by ANDing the filters together, and we never want to present the user with an option where they will hit a deadend. It is not clear there is a good way to let the user also OR in other filters or if that is beneficial. I think we probably do have a UX problem here where we should be suggesting alternative routes when the user has filtered too much to help them "dig out of the hole", but this isn't something we are going to fix now. The current behavior is the same as the existing JP Search (except we are missing a "Clear Filters" link).

**Comment by gibrown** [comment]
@jeherve we discussed and went over everything for this PR today. We think there are no ongoing blockers for it. I broke out a number of issues into separate bugs (and one enhancement request). Here is the full list:
- https://github.com/Automattic/jetpack/issues/13391
- https://github.com/Automattic/jetpack/issues/13808
- https://github.com/Automattic/jetpack/issues/13809
- https://github.com/Automattic/jetpack/issues/13810
- https://github.com/Automattic/jetpack/issues/13811
- https://github.com/Automattic/jetpack/issues/13812
- https://github.com/Automattic/jetpack/issues/13813
- https://github.com/Automattic/jetpack/issues/13814
- https://github.com/Automattic/jetpack/issues/13816

Some are corner cases that are unlikely to get fixed before the Release, but I have marked as high the ones we I think we should get fixed.

**Review by gibrown — DISMISSED** [review]
We've moved all of the corner cases into their own issues, so going to approve this for merging into master.

File: modules/search/instant-search/components/search-result-minimal.jsx
Comment (jeherve) [reply modules/search/instant-search/components/search-result-minimal.jsx]: To close the loop on this, this is tracked in #13810

File: modules/search/class.jetpack-search.php
Comment (jeherve) [reply modules/search/class.jetpack-search.php]: Marking this as resolved since this is being tracked and discussed in #13762

**Review by jeherve — DISMISSED** [review]
This looks good to me. I think we should be good to merge this into `master` for now.

I would recommend creating a new release branch based off `master` once this is merged, and rebasing / editing all your currently open PRs against that new branch:
#13774, #13796, #13350, #13349 (although this last one may now be closed in favor of #13774 maybe?)

*********

> I broke out a number of issues into separate bugs (and one enhancement request).

Thanks a lot for doing that! I think it makes the feedback on this whole big PR a bit more manageable. 

I had left one bit of feedback that I can't find among the open issues:

> If I hit my browser's return key after launching a search, the search query remains, but with one less character in the term I typed. I think the whole search query should disappear, don't you think?

It's possible you already addressed that comment of mine though, but I can't find it in here.

*******

>> Something seems off with the way tags are displayed now:
>> https://jeremy.hu/?post_tag=genericons
>
> This should be fixed.

I'm still seeing this for some posts though.

https://jeremy.hu/?s=bonobo

image

*********

The non-jetpack searchform built in Twenty Twenty seems to break when Instant Search is on. I get redirected to the homepage without any query string when I hit enter:

http://cl.ly/46f79ab7541a

Could it be that this was introduced in the last few changes to get the URL to behave when launching a search from a posts page?

**Comment by jsnmoon** [comment]
>>> Something seems off with the way tags are displayed now:
>>>jeremy.hu/?post_tag=genericons

>> This should be fixed.

> I'm still seeing this for some posts though.

It looks like we were omitting the additional styling applied to Gridicon icons; it should be fixed now.

**Review by gibrown — APPROVED** [review]
All good

**Pull request merged**
